pkg/agent/cache/manager.go

```diff
@@ -0,0 +1,382 @@
+	cancel           context.CancelFunc
+	doneCh           chan struct{}
+}
+
+func NewManager(ctx context.Context, c *MgrConfig) (Manager, error) {
+
+	cert, err := x509.ParseCertificate(c.BaseSVID)
+	if err != nil {
+		return nil, err
+	}
+	basespiffeID, err := uri.GetURINamesFromCertificate(cert)
+	if err != nil {
+		return nil, err
+	}
+	ctx, cancel := context.WithCancel(ctx)
+
+	return &manager{
+		managedCache:     NewCache(c.Logger),
+		serverCerts:      c.ServerCerts,
+		serverSPIFFEID:   c.ServerSPIFFEID,
+		serverAddr:       c.ServerAddr,
+		baseSVID:         c.BaseSVID,
+		baseSVIDKey:      c.BaseSVIDKey,
+		baseSPIFFEID:     basespiffeID[0],
+		baseRegEntries:   c.BaseRegEntries,
+		log:              c.Logger.WithField("subsystem_name", "cacheManager"),
+		spiffeIdEntryMap: make(map[string]CacheEntry),
+		entryRequestCh:   make(chan map[string][]EntryRequest),
+		regEntriesCh:     make(chan []*proto.RegistrationEntry),
+		cacheEntryCh:     make(chan CacheEntry),
+		ctx:              ctx,
+		cancel:           cancel,
+		doneCh:           make(chan struct{}),
+	}, nil
+}
+
+func (m *manager) Shutdown(err error) {
+	m.log.Debug("Shutting Down Cache Manager ", err)
+	m.reason = err
+	m.cancel()
+}
+func (m *manager) Done() <-chan struct{} {
+	return m.doneCh
+}
+func (m *manager) Err() error {
+	<-m.Done()
+	return m.reason
+}
+func (m *manager) Cache() Cache {
+	return m.managedCache
+}
+
+func (m *manager) Init() {
+	go func() {
+		defer close(m.doneCh)
+		var wg sync.WaitGroup
+
+		wg.Add(1)
+		go m.expiredCacheEntryHandler(5*time.Second, &wg)
+
+		wg.Add(1)
+		go m.regEntriesHandler(&wg)
+		m.log.Debug("Initializing Cache Manager")
+		m.regEntriesCh <- m.baseRegEntries
+		for {
+
+			var svid []byte
+			var key *ecdsa.PrivateKey
+			select {
+			case reqs := <-m.entryRequestCh:
+				for parentId, entryRequests := range reqs {
+					wg.Add(1)
+					if _, ok := m.spiffeIdEntryMap[parentId]; ok {
+						svid = m.spiffeIdEntryMap[parentId].SVID.SvidCert
+						key = m.spiffeIdEntryMap[parentId].PrivateKey
+					}
+					if parentId == m.baseSPIFFEID {
+						svid = m.baseSVID
+						key = m.baseSVIDKey
+					}
+					conn, err := m.getGRPCConn(svid, key)
+					if err != nil {
+						m.Shutdown(err)
+						break
+					}
+					go m.fetchSVID(entryRequests, node.NewNodeClient(conn), &wg)
+				}
+
+			case newCacheEntry := <-m.cacheEntryCh:
+				m.managedCache.SetEntry(newCacheEntry)
+				m.log.Debug("Updated Cache", "Cache:", newCacheEntry)
+
+			case <-m.ctx.Done():
+				m.log.Debug("Stopping Cache manager")
+				wg.Wait()
+				return
+			}
+		}
+	}()
+}
+
+func (m *manager) fetchSVID(requests []EntryRequest, nodeClient node.NodeClient, wg *sync.WaitGroup) {
+	defer wg.Done()
+	stream, err := nodeClient.FetchSVID(context.Background())
+	if err != nil {
+		m.Shutdown(err)
+		return
+	}
+	for _, req := range requests {
+
+		err := stream.Send(&node.FetchSVIDRequest{Csrs: append([][]byte{}, req.CSR)})
+
+		if err != nil {
+			m.Shutdown(err)
```

Comment:
`stream.Close()`

Reply:
Shutdown(err) can be called twice in the event that stream.CloseSend() in the defer returns non-nil error. Is that ok? Same thing with the Shutdown calls below